# Урок 2. Анализ датасета с помощью Pandas
## Условие 1: 1 задача
Скачать файл в уроке
Считать данные с помощью pandas
Вывести на экран первые 5 строк
Посмотреть на описание признаков и на их содержание


In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("kc-house-data (1).csv", encoding='windows-1251')
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

In [3]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

## Условие 2: 2 задача
Проведите первичный анализ данных

#### Изучите типы данных

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

#### Найдите количество пропущенных ячеек в данных

In [5]:
df.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

Ни в одном столбце нет пропущенных ячеек (Null).

#### Посчитайте основные статистики по всем признакам и поизучайте их

In [6]:
df.describe(include=('object', 'int64','float64'))

id             date         price      bedrooms  \
count   2.161300e+04            21613  2.161300e+04  21613.000000   
unique           NaN              372           NaN           NaN   
top              NaN  20140623T000000           NaN           NaN   
freq             NaN              142           NaN           NaN   
mean    4.580302e+09              NaN  5.400881e+05      3.370842   
std     2.876566e+09              NaN  3.671272e+05      0.930062   
min     1.000102e+06              NaN  7.500000e+04      0.000000   
25%     2.123049e+09              NaN  3.219500e+05      3.000000   
50%     3.904930e+09              NaN  4.500000e+05      3.000000   
75%     7.308900e+09              NaN  6.450000e+05      4.000000   
max     9.900000e+09              NaN  7.700000e+06     33.000000   

           bathrooms   sqft_living      sqft_lot        floors    waterfront  \
count   21613.000000  21613.000000  2.161300e+04  21613.000000  21613.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        2.114757   2079.899736  1.510697e+04      1.494309      0.007542   
std         0.770163    918.440897  4.142051e+04      0.539989      0.086517   
min         0.000000    290.000000  5.200000e+02      1.000000      0.000000   
25%         1.750000   1427.000000  5.040000e+03      1.000000      0.000000   
50%         2.250000   1910.000000  7.618000e+03      1.500000      0.000000   
75%         2.500000   2550.000000  1.068800e+04      2.000000      0.000000   
max         8.000000  13540.000000  1.651359e+06      3.500000      1.000000   

                view  ...         grade    sqft_above  sqft_basement  \
count   21613.000000  ...  21613.000000  21613.000000   21613.000000   
unique           NaN  ...           NaN           NaN            NaN   
top              NaN  ...           NaN           NaN            NaN   
freq             NaN  ...           NaN           NaN            NaN   
mean        0.234303  ...      7.656873   1788.390691     291.509045   
std         0.766318  ...      1.175459    828.090978     442.575043   
min         0.000000  ...      1.000000    290.000000       0.000000   
25%         0.000000  ...      7.000000   1190.000000       0.000000   
50%         0.000000  ...      7.000000   1560.000000       0.000000   
75%         0.000000  ...      8.000000   2210.000000     560.000000   
max         4.000000  ...     13.000000   9410.000000    4820.000000   

            yr_built  yr_renovated       zipcode           lat          long  \
count   21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean     1971.005136     84.402258  98077.939805     47.560053   -122.213896   
std        29.373411    401.679240     53.505026      0.138564      0.140828   
min      1900.000000      0.000000  98001.000000     47.155900   -122.519000   
25%      1951.000000      0.000000  98033.000000     47.471000   -122.328000   
50%      1975.000000      0.000000  98065.000000     47.571800   -122.230000   
75%      1997.000000      0.000000  98118.000000     47.678000   -122.125000   
max      2015.000000   2015.000000  98199.000000     47.777600   -121.315000   

        sqft_living15     sqft_lot15  
count    21613.000000   21613.000000  
unique            NaN            NaN  
top               NaN            NaN  
freq              NaN            NaN  
mean      1986.552492   12768.455652  
std        685.391304   27304.179631  
min        399.000000     651.000000  
25%       1490.000000    5100.000000  
50%       1840.000000    7620.000000  
75%       23

#### Пишите выводы

Выводы:

В наборе представлены данные о сделках с недвижимостью по 21 показателю.
Всего в наборе данные по 21 613 сделкам.
Большая часть данных представлена числовыми значениями (число комнат, площадь, цена и т.д.), есть дата сделки, координаты объекта.
Медианная сделка заключена по цене 4 500 У.Е. У объекта 3 спальни, 2.25 ванные комнаты, жилая площадб 1 910 кв.футов (~ 177 м.кв.) и т.д.

## Условие 3: 3 задача
Ответьте на несколько вопросов

#### 3.1 В каком диапазоне изменяются стоимости недвижимости?

In [7]:
df['price'].min(), df['price'].max()

(75000.0, 7700000.0)

#### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [8]:
living_sq = df['sqft_living'].sum()

In [9]:
total_sq = df['sqft_lot'].sum()

In [10]:
f"{round((living_sq / total_sq) * 100, 1)} %"

'13.8 %'

#### 3.3 Как много домов с разными этажами в данных?

In [11]:
unique_floors_value = set(df['floors'])

In [12]:
df['floors'].value_counts()

floors
1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: count, dtype: int64

#### 3.4 Насколько хорошие состояния у домов в данных?

In [13]:
df['condition'].value_counts()

condition
3    14031
4     5679
5     1701
2      172
1       30
Name: count, dtype: int64

#### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [14]:
df['yr_built'].min(), df['yr_built'].max()

(1900, 2015)

## Условие 4: 4 задача
Ответьте на несколько вопросов

#### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [15]:
round(df[df['bedrooms'] == 2].price.mean(), 2)

401372.68

#### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [16]:
round(df[df['price'] > 600000].sqft_lot.mean(), 2)

20442.52

#### 4.3 Как много домов коснулся ремонт?

In [17]:
df[df['yr_renovated'] != 0].value_counts().sum()


914

#### 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [18]:
real_estate_price_10 = round(df[df['grade'] > 10].price.mean(), 2)

In [19]:
real_estate_price_4 = round(df[df['grade'] < 4].price.mean(), 2)

In [20]:
difference_10_4 = real_estate_price_10 - real_estate_price_4
difference_10_4

1488885.12

## Условие 5: 5 задача
Ответьте на несколько вопросов

#### 5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [21]:
df[
    (df['waterfront'] != 0) &
    (df['bathrooms'] >= 3) &
    (df['sqft_basement'] != 0)
].shape[0]

41

#### 5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [22]:
df_fream_house = df[
    (df['yr_built'] >= 1980) &
    (df['condition'] == 5) &
    (df['view'] != 0) |
    (df['yr_built'] >= 1980) &
    (df['condition'] == 5) &
    (df['waterfront'] != 0)
]

In [23]:
df_fream_house['price'].min(), df_fream_house['price'].max()

(415000.0, 3000000.0)

#### 5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [24]:
df[
    (df['sqft_basement'] == 0) &
    (df['floors'] >= 2) &
    (df['price'] < 150000)
].grade.mean()

7.0